### DISCUSSION TITLE
what can be done to release CSharpCompilation from memory ?

### DISCUSSION THREAD

**Discussion by birojnayak** [dotnet/roslyn#66828] [General]
CSharpCompilation always stays in memory even I am done with compilation unit. Why this doesn't get Disposed ? Or What am I missing here ?

The API that I use to get the compilation is Project.GetCompilationAsync().

<img width="1232" alt="memoryusage" src="https://user-images.githubusercontent.com/49173255/218299341-31317127-9a01-4cc6-9e02-ea0927c50dee.png">

**Comment by CyrusNajmabadi** [comment]
It doesn't get disposed because something else may try to get the compilation from that project. And throwing out away to create it again would be extremely expensive. 

Note, this is the normal behavior for the IDE. We have hundreds of services that operate on their own cadence that need this compilation.  Throwing it away wood just lead to much more cpu and memory churn.

**Reply by birojnayak** [reply]
@CyrusNajmabadi thanks for prompt response appreciate. I create an AdHocWorkSpace , add multiple projects and build each project and get compilation unit and syntax tree. Once I am done, I dispose the AdHocWorkSpace, I explicitly get metareference for each project and dispose them, but somehow I can't get hold of the CSharpCompilation to dispose. I am using roslyn as more of API's not from Visual Studio. Let me know if there is a way completely throw them out to release memory pressure. Because next time I would be dealing with next Solution with new projects. Do you think ALC would help ? Or let me know if there is a better way.

**Reply by CyrusNajmabadi** [reply]
> but somehow I can't get hold of the CSharpCompilation to dispose.

Compilations are not disposable.  There is no concept of disposing htem.  

> Let me know if there is a way completely throw them out to release memory pressure.

Try setting the workspace's current solution to point at something else (like an empty solution).

if that doesn't help, please provide a memory diagnostic that shows what is still rooting the compilations.  thanks!

**Reply by birojnayak** [reply]
@CyrusNajmabadi here are some weakreference.... pointing to a new solution creates and holds on to another CShapCompilation... 
<img width="1080" alt="Screenshot 2023-02-15 at 9 32 18 AM" src="https://user-images.githubusercontent.com/49173255/219107779-1a566c48-9ac9-4f7a-b55f-d83c7992cce8.png">
<img width="1076" alt="Screenshot 2023-02-15 at 9 32 06 AM" src="https://user-images.githubusercontent.com/49173255/219107796-99d84458-6852-4202-a1d7-fb0a32c9023e.png">
<img width="1073" alt="Screenshot 2023-02-15 at 9 31 53 AM" src="https://user-images.githubusercontent.com/49173255/219107806-d5f64d61-8bee-4bb7-a0b3-54e26e4e9f8a.png">
<img width="1081" alt="Screenshot 2023-02-15 at 9 31 34 AM" src="https://user-images.githubusercontent.com/49173255/219107810-0d41acd3-8f34-4a9d-ac07-b5d78b1c16c7.png">

**Reply by CyrusNajmabadi** [reply]
Weak refs should not keep anything alive. Looks like the GC is just deciding it doesn't need to let these go yet. Not much roslyn can do about that.

**Reply by jasonmalinowski** [reply]
Also those screenshots are in many cases static ConditionalWeakTables -- there will always be one of them; the question is whether anything is keeping those keys alive. I'm not sure if the VS memory handles them well -- this might be a case where we have to (unfortunately) pull WinDBG out.

**Reply by birojnayak** [reply]
@jasonmalinowski @CyrusNajmabadi let me send you guys an email while I set up windbg/sos do more digging.
